loadbalancer/algorithm.go

```diff
@@ -97,25 +101,53 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, w []int, choice int) rou
 		ctx.Route.LBEndpoints[choice].Detected,
 	)
 
-	if rnd.Float64() < f {
+	if sr.getFloat64() < f {
 		return ctx.Route.LBEndpoints[choice]
 	}
 
-	return shiftToRemaining(rnd, ctx, w, now)
+	return shiftToRemaining(ctx, sr, w, now)
+}
+
+// safeRand is a struct for managing concurrent safe random access
+// with the custom source.
+type safeRand struct {
+	rnd *rand.Rand
+	mx  sync.Mutex
+}
+
+func newSafeRand() *safeRand {
+	src := rand.NewSource(time.Now().UnixNano())
+	return &safeRand{
+		rnd: rand.New(src),
+	}
+}
+
+// getIntn returns random int within the given range.
+func (s *safeRand) getIntn(lbe int) int {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+	return s.rnd.Intn(lbe)
+}
+
+// getFloat64 returns random float64
+func (s *safeRand) getFloat64() float64 {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+	return s.rnd.Float64()
 }
 
 type roundRobin struct {
 	mx         sync.Mutex
 	index      int
-	rnd        *rand.Rand
+	sr         *safeRand
 	fadeInCalc []int
 }
 
 func newRoundRobin(endpoints []string) routing.LBAlgorithm {
-	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	sr := newSafeRand()
```

Comment:
`roundRobin` already has mutex that guards both index update and random generation in `Apply` and it is safe to call it without mutex in constructor, so it is not needed here.